### ISSUE TITLE
Filter MultiAutocomplete now adds every intermediate value as a filter

### ISSUE THREAD

**Issue by willbryant** [metabase/metabase#43426]
### Describe the bug

When typing a value into a MultiAutocomplete filter field, a filter is added not just for what you actually select, but for each of the values that the autocomplete control had after typing each letter. Please see the attached video for a demo.

The same problem happens if you choose some values from the MultiAutocomplete drop-down and then X them out; they're still added as filters.

This only happens for certain fields, and I'm not sure why - I haven't found the right Sample Database fields to reproduce it. It doesn't happen when filtering on that table directly, I think - it needs to come from an association. But some other similar name fields on associated tables don't trigger the bug, and I can't see why they're different.

This is a regression in 0.49. I've done a git bisect and the problem was introduced in commit e64daa4f452af197d7d3f62bac083fecf482b86a which is part of the improvements @ranquild made in #40278.

### To Reproduce

1. Go to a table that exhibits the problem
2. Click on 'Filter'
3. Scroll down to the associated table that exhibits the problem
4. Go to an auto-complete-able text field and type a few letters
5. Select a value
6. Apply the filter
7. Observe that a number of filters have been added, one after each character typed

### Expected behavior

Only the selected values shown in the autocomplete should be added as a filter.

### Logs

No JS logs emitted while the frontend is hitting this bug. Nothing interesting I can see on the backend.

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "21.0.3",
    "java.vendor": "Homebrew",
    "java.vendor.url": "https://github.com/Homebrew/homebrew-core/issues",
    "java.version": "21.0.3",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "21.0.3",
    "os.name": "Mac OS X",
    "os.version": "14.4.1",
    "user.language": "en",
    "user.timezone": "Pacific/Auckland"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "bigquery-cloud-sdk"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "2.1.214 (2022-06-13)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "2.1.214 (2022-06-13)"
      }
    },
    "run-mode": "dev",
    "version": {
      "date": "2022-09-06",
      "src_hash": "8a88eaab9b69213780bd2547d41bd97f1f2e9855",
      "tag": "v0.45.0-SNAPSHOT",
      "branch": "master",
      "hash": "93d3df5"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

Medium

### Additional context

Demo from our test environment:

https://github.com/metabase/metabase/assets/12062/31033483-ae4b-4bbd-a872-c51b866a42b0

**Comment by willbryant** [comment]
I think I've figured out why this only happens on some fields. First of all, this only happens on the filter modal, not the editor screen. Secondly, it only happens when you see "Search the list" fields.

I believe this is because it's only in the case where FilterValuePicker returns a ListValuePicker and where ListValuePicker returns an AutocompletePicker rather than a CheckboxGridPicker.

The filter dialog doesn't experience this issue because in the !props.compact case. And similarly, a lot fields won't exhibit this bug because in the props.fieldValues.length > MAX_INLINE_OPTIONS case it will return CheckboxGridPicker.

And you won't get the ListValuePicker ("Search the list") unless canListFieldValues, which looks to me like it's checking mainly that the list of values is known and not too long (!has_more_values).

Therefore, to repro this bug, you need some distinct data with more than MAX_INLINE_OPTIONS (12) values, but not so many that you get into the !has_more_values case, I think?

**Comment by willbryant** [comment]
I thought it might be this part of the commit:

```
@@ -71,6 +71,7 @@ function FilterValuePicker({
         fieldValues={fieldData.values}
         selectedValues={selectedValues}
         placeholder={t`Search the list`}
+        shouldCreate={shouldCreate}
         autoFocus={autoFocus}
         compact={compact}
         onChange={onChange}
```

Because that's the only one that *added* `shouldCreate` rather than *replacing* a `canAddValue` with `shouldCreate`. But reverting that line doesn't seem to fix it. So I think it's something to do with the onChange behavior in MultiAutocomplete?

**Comment by ranquild** [comment]
Repro with Sample Database:
- Admin -> Table Metadata -> Products -> Title -> Change "Filtering on this field" to "Search box"
- New -> Question -> Orders
- Join Orders on ID = ID (self join)
- Visualize
- Filter -> See multiple 2 join groups for Product because there are 2 PRODUCT_ID now
- Select the second Product group
- Change operator for Title to be `Is`, start typing some value, then select it from the dropdown
- Click Apply filters
- See multiple filters created
- This is because we lose track of implicitly joined fields where there is more than 1 `source-field`

Demo https://www.loom.com/share/cd8eb83f6de84f7289e36af33696f92a
I'll have a fix shortly that would remove some of the symptoms - not fully fix this issue.

**Comment by ranquild** [comment]
@willbryant it's partially fixed in master now and will be released in some 0.50.x release soon. I'll leave this issue open as it's still possible to face it. Reassigning to another team because there is an internal issue with implicit field references that I cannot fix here.

**Comment by ranquild** [comment]
Closing this issue (by https://github.com/metabase/metabase/pull/45532). I'll create another one for other issues with implicit fields.
